File: src/apps/EventViewer/src/FrontmostApplicationController.m
#import "FrontmostApplicationController.h"
#import "libkrbn/libkrbn.h"
#import <pqrs/weakify.h>

@interface FrontmostApplicationController ()

@property NSMutableAttributedString* text;
@property(unsafe_unretained) IBOutlet NSTextView* textView;

- (void)callback:(NSString*)bundleIdentifier
        filePath:(NSString*)filePath;

@end

static void staticCallback(const char* bundle_identifier,
                           const char* file_path,
                           void* context) {
  FrontmostApplicationController* p = (__bridge FrontmostApplicationController*)(context);
  [p callback:[NSString stringWithUTF8String:bundle_identifier]
      filePath:[NSString stringWithUTF8String:file_path]];
}

@implementation FrontmostApplicationController

- (void)setup {
  self.text = [NSMutableAttributedString new];

  libkrbn_enable_frontmost_application_monitor(staticCallback,
                                               (__bridge void*)(self));
}

- (void)dealloc {
  libkrbn_disable_frontmost_application_monitor();
}

- (void)callback:(NSString*)bundleIdentifier filePath:(NSString*)filePath {
  @weakify(self);
  dispatch_async(dispatch_get_main_queue(), ^{
    @strongify(self);
    if (!self) {
      return;
    }

    // Update self.text

    if (![@"org.pqrs.Karabiner.EventViewer" isEqualToString:bundleIdentifier] &&
        ![@"org.pqrs.Karabiner-EventViewer" isEqualToString:bundleIdentifier]) {
      // Clear if text is huge.
      if (self.text.length > 4 * 1024) {
        [self.text setAttributedString:[[NSAttributedString alloc] initWithString:@""]];
      }

      NSString* bundleIdentifierLine = [NSString stringWithFormat:@"Bundle Identifier:  %@\n", bundleIdentifier];
      NSString* filePathLine = [NSString stringWithFormat:@"File Path:          %@\n\n", filePath];
      NSDictionary* attributes = @{
        NSForegroundColorAttributeName : NSColor.textColor,
        NSFontAttributeName : [NSFont fontWithName:@"Menlo" size:11],
      };

      [self.text appendAttributedString:[[NSAttributedString alloc] initWithString:bundleIdentifierLine
                                                                        attributes:attributes]];
      [self.text appendAttributedString:[[NSAttributedString alloc] initWithString:filePathLine
                                                                        attributes:attributes]];
    }

    // Update self.textView

    NSTextStorage* textStorage = self.textView.textStorage;
    [textStorage beginEditing];
    if (self.text.length == 0) {
      NSString* placeholder = @"Please switch to apps which you want to know Bundle Identifier.";
      [textStorage setAttributedString:[[NSAttributedString alloc] initWithString:placeholder]];
    } else {
      [textStorage setAttributedString:self.text];
    }
    [textStorage endEditing];

    [self.textView scrollRangeToVisible:NSMakeRange(self.textView.string.length, 0)];
  });
}

@end

File: src/apps/EventViewer/src/FrontmostApplicationController.swift
import Cocoa

private func callback(_ bundleIdentifier: UnsafePointer<Int8>?,
                      _ filePath: UnsafePointer<Int8>?,
                      _ context: UnsafeMutableRawPointer?)
{
  if bundleIdentifier == nil { return }
  if filePath == nil { return }

  let identifier = String(cString: bundleIdentifier!)
  let path = String(cString: filePath!)
  let obj: FrontmostApplicationController? = unsafeBitCast(context, to: FrontmostApplicationController.self)

  if identifier == "org.pqrs.Karabiner-EventViewer" { return }

  // Update obj.text

  guard let font = NSFont(name: "Menlo", size: 11) else { return }
  let attributes = [
    NSAttributedString.Key.font: font,
    NSAttributedString.Key.foregroundColor: NSColor.textColor,
  ]

  if obj!.text.length > 4 * 1024 {
    obj!.text.setAttributedString(NSAttributedString(
      string: "", attributes: attributes
    ))
  }

  obj!.text.append(NSAttributedString(
    string: "Bundle Identifier:  \(identifier)\n",
    attributes: attributes
  ))
  obj!.text.append(NSAttributedString(
    string: "File Path:          \(path)\n\n",
    attributes: attributes
  ))

  // Update obj.textView

  DispatchQueue.main.async { [weak obj] in
    guard let obj = obj else { return }

    guard let textStorage = obj.textView!.textStorage else { return }

    textStorage.beginEditing()
    textStorage.setAttributedString(obj.text)
    textStorage.endEditing()

    obj.textView!.scrollRangeToVisible(NSMakeRange(obj.text.length, 0))
  }
}

@objc
public class FrontmostApplicationController: NSObject {
  @IBOutlet var textView: NSTextView?
  let text: NSMutableAttributedString
  let attributes: [NSAttributedString.Key: Any]

  override init() {
    text = NSMutableAttributedString()

    var attributes: [NSAttributedString.Key: Any] = [
      NSAttributedString.Key.foregroundColor: NSColor.textColor,
    ]
    if let font = NSFont(name: "Menlo", size: 11) {
      attributes[NSAttributedString.Key.font] = font
    }
    self.attributes = attributes

    super.init()
  }

  deinit {
    libkrbn_disable_frontmost_application_monitor()
  }

  @objc
  public func setup() {
    let obj = unsafeBitCast(self, to: UnsafeMutableRawPointer.self)
    libkrbn_enable_frontmost_application_monitor(callback, obj)

    let textStorage = textView?.textStorage
    textStorage?.beginEditing()
    textStorage?.setAttributedString(NSAttributedString(
      string: "Please switch to apps which you want to know Bundle Identifier.",
      attributes: attributes
    ))
    textStorage?.endEditing()
  }
}
